### PR TITLE
add more graph methods

### PR DESCRIPTION
**Pull request by juanvallejo**
Prerequisite to https://github.com/openshift/origin/pull/18466

cc @deads2k

### CODE REVIEW

File: tools/depcheck/pkg/cmd/util/filter.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package trace
+package util
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/util/filter.go]: is util a new package.  I hate "bucket-o-stuff" packages.

File: tools/depcheck/pkg/cmd/util/pkg.go [hunk 2]
```diff
@@ -91,7 +95,10 @@ func BuildGraph(packages *PackageList, excludes []string) (*depgraph.MutableDire
 
 			to, exists := g.NodeByName(dependency)
 			if !exists {
-				return nil, fmt.Errorf("expected child node for dependency %q was not found in graph", dependency)
+				// if a package imports a dependency that we did not visit
+				// while traversing the code tree, ignore it, as it is not
+				// required for the root repository to build.
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/util/pkg.go]: glog here as I recall

File: tools/depcheck/pkg/cmd/util/pkg.go [hunk 3]
```diff
@@ -108,6 +115,43 @@ func BuildGraph(packages *PackageList, excludes []string) (*depgraph.MutableDire
 	return g, nil
 }
 
+// CalculatePackages receives a set of go import paths and runs "go list"
+// using each path as an entrypoint.
+// Returns a PackageList containing dependency and importPath data for each package.
+func CalculatePackages(entrypoints []string) (*PackageList, error) {
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/util/pkg.go]: This looks like it is translate user input into runtime objects.  It belongs close to your command.

File: tools/depcheck/pkg/graph/graph.go [hunk 4]
```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+		if !ok {
+			continue
+		}
+		if len(g.To(n)) > 0 {
+			continue
+		}
+
+		skip := false
+		for _, toIgnore := range ignore {
+			if strings.HasPrefix(node.UniqueName, toIgnore) {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: No prefix matching.  explode the list before you get here.

File: tools/depcheck/pkg/graph/graph.go [hunk 5]
```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+		node, ok := n.(*Node)
+		if !ok {
+			continue
+		}
+		if len(g.To(n)) > 0 {
+			continue
+		}
+
+		skip := false
+		for _, toIgnore := range ignore {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: make a private, `contains(needle string, haystack []string)` so this doesn't look as ugly.

File: tools/depcheck/pkg/graph/graph.go [hunk 6]
```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+		return []*Node{}
+	}
+
+	return append(removed, g.PruneOrphans(ignore)...)
+}
+
+// FindExclusiveDependencies receives a set of targetNodes.
+// The graph is cloned, and all targetNodes are removed from it.
+// Returns the resulting set of orphaned nodes recursively pruned from the cloned graph.
+func (g *MutableDirectedGraph) FindExclusiveDependencies(targetNodes []*Node, ignore []string) []*Node {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: I don't think this should be base functionality of your graph.  Move this to an analysis package.

File: tools/depcheck/pkg/graph/graph.go [hunk 7]
```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+	for _, target := range targetNodes {
+		newGraph.RemoveNode(target)
+	}
+
+	return newGraph.PruneOrphans(ignore)
+}
+
+// HasPathFromTo receives two nodes and determines if node B can be reached from node A.
+// Returns false if A and B are the same node.
+func (g *MutableDirectedGraph) HasPathFromTo(A, B graph.Node) bool {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: You really shouldn't have this function.  The base graph has functionality for this.  Also, I'd hope this is dead.

File: tools/depcheck/pkg/graph/graph.go [hunk 8]
```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+
+		for _, inbound := range g.To(n) {
+			unseen = append(unseen, inbound)
+		}
+	}
+
+	return false
+}
+
+func (g *MutableDirectedGraph) Copy() *MutableDirectedGraph {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: godoc noting each node is a shallow copy

File: tools/depcheck/pkg/graph/graph.go [hunk 9]
```diff
@@ -55,7 +60,140 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+
 func (g *MutableDirectedGraph) NodeByName(name string) (graph.Node, bool) {
 	n, exists := g.nodesByName[name]
 	return n, exists && g.DirectedGraph.Has(n)
 }
+
+// PruneOrphans recursively removes nodes with no inbound edges.
+// Receives a list of UniqueName prefixes of nodes to ignore.
+// Returns a list of recursively pruned nodes.
+func (g *MutableDirectedGraph) PruneOrphans(ignore []string) []*Node {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: needs a test

Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: I think that `ignore` is a base quality of the graph.  `originatingNodes` perhaps?

File: tools/depcheck/pkg/cmd/util/pkg.go (hunk 3 above)
Comment (deads2k) [inline tools/depcheck/pkg/cmd/util/pkg.go]: Whereever this ends up. Unit test

Comment (deads2k) [reply tools/depcheck/pkg/cmd/util/pkg.go]: > Whereever this ends up. Unit test

Unit test if at all possible I guess.

**Comment by juanvallejo** [comment]
@deads2k thanks, review comments addressed

**Comment by juanvallejo** [comment]
@deads2k not sure that the test for `CalculatePackages` is going to work:
```
warning: ignoring symlink /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin
warning: "github.com/openshift/origin/tools/depcheck/test/testrepo/pkg/..." matched no packages
warning: ignoring symlink /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin
warning: "github.com/openshift/origin/tools/depcheck/test/testrepo/cmd/..." matched no packages
--- FAIL: TestCalculatePackages (0.06s)
	pkg_test.go:271: package mismatch: expected 3 packages to have been generated, saw 0
FAIL
```

I could wire a "CommandRunner" through to the `CalculatePackages` method. Would allow me to fake go list output for the test

**Comment by deads2k** [comment]
> @deads2k not sure that the test for CalculatePackages is going to work:

Ok.  I'm willing to forgo it for now.

File: tools/depcheck/pkg/cmd/pkg.go [hunk 10]
```diff
@@ -108,6 +118,43 @@ func BuildGraph(packages *PackageList, excludes []string) (*depgraph.MutableDire
 	return g, nil
 }
 
+// CalculatePackages receives a set of go import paths and runs "go list"
+// using each path as an entrypoint.
+// Returns a PackageList containing dependency and importPath data for each package.
+func CalculatePackages(entrypoints []string) (*PackageList, error) {
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/pkg.go]: uncalled in this pull?

File: tools/depcheck/pkg/cmd/pkg.go [hunk 11]
```diff
@@ -108,6 +118,43 @@ func BuildGraph(packages *PackageList, excludes []string) (*depgraph.MutableDire
...
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			}
+
+			list.Add(pkg)
+		}
+	}(pkgs)
+
+	// no need to wait for goroutine to finish, golist.Run() will
+	// block until the command finishes executing.
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/pkg.go]: Doesn't seem likely.  Wouldn't it wait until the buffer was consumed, not until the list.Add call finished?

File: tools/depcheck/pkg/graph/graph.go [hunk 12]
```diff
@@ -55,7 +59,99 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+
 func (g *MutableDirectedGraph) NodeByName(name string) (graph.Node, bool) {
 	n, exists := g.nodesByName[name]
 	return n, exists && g.DirectedGraph.Has(n)
 }
+
+// PruneOrphans recursively removes nodes with no inbound edges.
+// Receives a list of UniqueNames of nodes to ignore.
+// Returns a list of recursively pruned nodes.
+func (g *MutableDirectedGraph) PruneOrphans(originatingNodes []string) []*Node {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: any reason originatingNodes can't be on `g`?

File: tools/depcheck/pkg/graph/graph.go [hunk 13]
```diff
@@ -55,7 +59,99 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+	newGraph := NewMutableDirectedGraph()
+
+	// copy nodes
+	for _, n := range g.Nodes() {
+		newNode, ok := n.(*Node)
+		if !ok {
+			continue
+		}
+
+		// ignore error: the only error that could happen is if a name collision occurs.
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: panic instead of ignore

File: tools/depcheck/pkg/graph/graph.go (hunk 12 above)
Comment (juanvallejo) [reply tools/depcheck/pkg/graph/graph.go]: > any reason originatingNodes can't be on g?

That sounds a bit better. Will update

File: tools/depcheck/pkg/cmd/pkg.go (hunk 10 above)
Comment (juanvallejo) [reply tools/depcheck/pkg/cmd/pkg.go]: yeah, will be called by the trace and analyze commands.
Can wait to add this as part of the trace cmd if that makes more sense

**Comment by juanvallejo** [comment]
@deads2k thanks, review comments addressed

File: tools/depcheck/pkg/cmd/pkg.go [hunk 14]
```diff
@@ -1,10 +1,12 @@
-package trace
+package cmd
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/pkg.go]: why isn't this part of tools/depcheck/pkg/graph?

File: tools/depcheck/pkg/cmd/pkg_test.go [hunk 15]
```diff
@@ -1,4 +1,4 @@
-package trace
+package cmd
```
Comment (deads2k) [inline tools/depcheck/pkg/cmd/pkg_test.go]: you're picking up on bad patterns in the main code.  package "cmd" is a bad idea, not a good one.  These are all related to the graph, just not methods on it.  You could co-locate

File: tools/depcheck/pkg/graph/graph.go [hunk 16]
```diff
@@ -55,7 +60,116 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
...
+// removing all of its outbound and inbound edges as well.
+func (g *MutableDirectedGraph) RemoveNode(n *Node) {
+	delete(g.nodesByName, n.UniqueName)
+	g.DirectedGraph.RemoveNode(n)
+}
+
+// SetRoots receives a set of node names and adds the
+// corresponding nodes to a list of known roots.
+// Returns an error if at least one given name does not correspond to a node
+func (g *MutableDirectedGraph) SetRoots(nodeNames []string) error {
```
Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: Seems weird.  Why don't we know this ahead of time?

Comment (deads2k) [inline tools/depcheck/pkg/graph/graph.go]: only called from a test?

File: tools/depcheck/test/testrepo/cmd/cmd.go [hunk 17]
```diff
@@ -0,0 +1,9 @@
+package main
```
Comment (deads2k) [inline tools/depcheck/test/testrepo/cmd/cmd.go]: This package is now dead, correct?

**Comment by deads2k** [comment]
minor comments, lgtm otehrwise

**Comment by juanvallejo** [comment]
@deads2k thanks, review comments addressed

**Pull request merged**